Quick Start Guide for the New Posit Environment (#138)
Public-Health-Scotland/technical-docs: Posit Infrastructure/New Posit Environment Quick Start Guide.md

# Quick Start Guide: Transition to the New Posit Environment

## 1. Overview

### Purpose
This Quick Start Guide is designed to help you transition smoothly to the new Posit environment. It provides essential information and step-by-step instructions to get you started on day one.  It is not a comprehensive guide; further documentation and guidance will be added to the [PHS Data Science Knowledge Base](https://public-health-scotland.github.io/knowledge-base/) in due course.

### Audience
This guide is intended for existing users of the old Posit environment who are moving to the new Posit environment.

### Scope
This guide covers:
- A high-level overview of the new Posit environment's specifications and features.
- Step-by-step instructions to get you started with logging in and setting up your environment.
- Details on how to get help.
- Links to other useful information and resources.

---

## 2. Environment Specifications

### Operating System and Application Specifications

> - **Operating Sytem**

Reviewer: ```suggestion
- **Operating System**
```

- R Packages that require C++14 or C++17 support, such as `{ranger}`, now install without error.

---

## 3. URLs for New Posit Applications
- **Posit Workbench**: https://pwb-prod.publichealthscotland.org/
- **Posit Package Manager**: https://ppm-prod.publichealthscotland.org/
- **Posit Connect**: _Coming soon..._

## 4. Logging in and Starting a Session

- Follow the link above to the new production Posit Workbench environment and log in using your LDAP username and password.  These are the same credentials you have used for the old Posit Workbench environment.
- The "New Session" dialog window has changed slightly.  On first requesting a new session, you are given a choice of the four available IDEs:

![New Session Dialog 1](https://github.com/user-attachments/assets/000c2943-bd4b-4818-974a-679dbb8987a9)

- As before, you need to enter the number of CPUs and Memory you need for your session.  **Important Change: you must specify the amount of memory you require in gigabytes (GB)**:

![New Session Dialog 2](https://github.com/user-attachments/assets/bd593345-4b9a-49a4-a4c4-f52008524438)

> **Remember to select the minimum number of CPUs and amount of Memory required for the work you intend to do.**  In most cases, you only require more than 1 CPU if you are knowingly going to be running code using multiple threads or parallel processing.  The amount of Memory your require is dependent on the number of R packages you intend to use and the size of the dataset you will be working with.  There is further guidance available on the PHS Data Science Knowledge Base at https://public-health-scotland.github.io/knowledge-base/docs/Posit%20Infrastructure?doc=Memory%20Usage%20in%20SMR01.md

Reviewer 2: "The amount of Memory your require" should be "you require"

> **Remember to select the minimum number of CPUs and amount of Memory required for the work you intend to do.**  In most cases, you only require more than 1 CPU if you are knowingly going to be running code using multiple threads or parallel processing.  The amount of Memory your require is dependent on the number of R packages you intend to use and the size of the dataset you will be working with.  There is further guidance available on the PHS Data Science Knowledge Base at https://public-health-scotland.github.io/knowledge-base/docs/Posit%20Infrastructure?doc=Memory%20Usage%20in%20SMR01.md

Reviewer: ```suggestion
**Remember to select the minimum number of CPUs and amount of Memory required for the work you intend to do.**  In most cases, you only require more than 1 CPU if you are knowingly going to be running code using multiple threads or parallel processing. The amount of memory you require is dependent on the number of R packages you intend to use and the size of the dataset you will be working with.  There is further guidance available on the PHS Data Science Knowledge Base at https://public-health-scotland.github.io/knowledge-base/docs/Posit%20Infrastructure?doc=Memory%20Usage%20in%20SMR01.md
```

Also is this really correct?
> The amount of memory you require is dependent on **the number of R packages you intend to use**

Author: ```suggestion
**Remember to select the minimum number of CPUs and amount of memory required for the work you intend to do.**  In most cases, you only require more than 1 CPU if you are knowingly going to be running code using multiple threads or parallel processing.  The amount of memory you require is primarily dependent on the size of the dataset you will be working with.  There is further guidance available on the PHS Data Science Knowledge Base at https://public-health-scotland.github.io/knowledge-base/docs/Posit%20Infrastructure?doc=Memory%20Usage%20in%20SMR01.md
```

The number of R packages you intend to use will have an impact, but the primary impact on memory is the size of dataset, so I've updated the text to reflect this.  Does it read ok @Moohan ?

![Starting Session Dialog](https://github.com/user-attachments/assets/2966723d-9b0a-42e2-99cb-9b6a8c1aa6e7)

**Important Note: normally sessions will take just a few seconds to start, but if the Kubernetes cluster needs to scale up to accommodate your session, it can take up to 10 minutes for your session to start.  This is normal and expected behaviour.**

- The session will be listed on the Posit Workbench homepage as "PENDING" during this process:

![RStudio Pro Session Pending](https://github.com/user-attachments/assets/0713f025-c984-4a46-9c48-bbbf2fb1ea35)

- If the "Join session when ready" checkbox was ticked in the "New Session" dialog window, the session will automatically load in your browser.  Alternatively, once the session shows as "ACTIVE" on the Posit Workbench homepage, you can click the session to open it:

![RStudio Pro Session Active](https://github.com/user-attachments/assets/e7e81280-e71b-4d5e-9742-4ca2650bb9b3)

## 5. Configuring your RStudio Pro environment

### Global Options

- Once your RStudio Pro session has started, please ensure that you configure your settings as follows:

>   - Access the RStudio Global Opations menu by going to _Tools > Global Options..._

Reviewer: ```suggestion
  - Access the RStudio Global Options menu by going to _Tools > Global Options..._
```

  - In the General menu, be sure that the settings match those in the screenshot below:

![Global Options](https://github.com/user-attachments/assets/c1441dba-b6af-4bec-b017-cc371bd14091)

  - _Optional, but recommended:_ On the Appearance menu, change the editor font to "Cascadia Code".  This font has support for ligatures that map multiple characters to single glyphs e.g. rendering "<-" as a single arrow glyph.

![Cascadia Code](https://github.com/user-attachments/assets/591ad6c5-4fad-4c25-9142-4ae0df288428)

### Installing Packages

> - You will need to install R packages that you need.  These will not transfer over from the old Posit environment.

Reviewer: ```suggestion
- You will need to install the R packages that you need.  These will not transfer over from the old Posit environment.
```

> - R (in both RStudio Pro and Visual Studio Code) has been configured to install binary versions of packages by default from Posit Package Manager.  This significantly reduces the installation time for packages.

Reviewer: ```suggestion
- R (in both RStudio Pro and Visual Studio Code) has been configured to install binary versions of packages by default from the PHS Posit Package Manager.  This significantly reduces the installation time for packages.
```

- You can further speed up package installation time by requesting more than 1 CPU for your session and running the following code before installing your packages:

```r
install.packages("parallelly")
available_cores <- as.numeric(parallelly::availableCores())
options(Ncpus = available_cores)
Sys.setenv(MAKEFLAGS = paste("-j", as.character(available_cores), sep = ""))
```

This will force the R packages to install in parallel, if possible, and any packages that do not have a pre-compiled binary will have their source code compiled in parallel (i.e. faster).

> - On day one, if you need to install our PHS R packages (`{phsmethods}`, `{phsopendata}`, `{phstemplates}`, `{phsstyles}` and `{slfhelper}`), you will have to install these directly from GitHub using the `remotes::install_github()` function.  In a few weeks time, the Data Science Team will configure Posit Package Manager with pre-built binaries for these packages.

Reviewer: ```suggestion
- On day one, if you need to install our PHS R packages (`{phsmethods}`, `{phsopendata}`, `{phstemplates}`, `{phsstyles}`, `{phsverse}` and `{slfhelper}`), you will have to install these directly from GitHub using the `remotes::install_github()` function.  In a few weeks time, the Data Science Team will configure Posit Package Manager with pre-built binaries for these packages.
```

---

## 6. Getting Support

**Important Change - please read carefully**

There is a new support process for the new Posit environment as follows:

1. All requests for support and to report issues with the new Posit environment must be made via the form at https://forms.office.com/e/pqmmNMkhT6

These requests for support will first reach the Data Science Team, who will provide first line support where possible.  If the Data Science Team cannot resolve the issue, it is passed to second-line support.

> 2. Second line support for the new Posit environment will be provided by Jumping Rivers.  An Engineer or Data Scientist at Jumping Rivers may contact you directly about the issue you are experiencing.

Reviewer: ```suggestion
2. Second-line support for the new Posit environment will be provided by Jumping Rivers.  An Engineer or Data Scientist at Jumping Rivers may contact you directly about the issue you are experiencing.
```

> 3. NSS DaS will only become involved in exceptional circumstances, such as a network outage, or if a significant change is required in the Microsoft Azure cloud computing environment.  The Data Science Team will co-ordinate the communication between Jumping Rivers and NSS DaS in these instances.

Reviewer: ```suggestion
3. NSS DaS will only become involved in exceptional circumstances, such as a network outage, or if a significant change is required in the Microsoft Azure cloud computing environment.  The Data Science Team will coordinate the communication between Jumping Rivers and NSS DaS in these instances.
```

**Important Notes**

> - You must complete the form linked to above in order to receive support relating to the new Posit environment.

Reviewer: ```suggestion
- You must complete the form linked above to receive support relating to the new Posit environment.
```

- Do not contact Jumping Rivers, unless they request further information from you directly.
- Do not raise a request or incident in Service Now, unless directed to do so by the Data Science Team.

---

## 5. FAQs
### What if I encounter a problem?
- Contact the Data Science Team for support using the form at [https://forms.office.com/e/pqmmNMkhT6]
### How can I get access to Posit Connect?

> - Posit Connect has been deployed into the new Posit environment along with Posit Workbench and Posit Package Manager.  Work continues to further test the capabilities of Posit Connect and assess how we can best make use of this application in PHS.  If you would like to get involved in this testing, please contact ...

Reviewer 2: Ross Elder at Ross.Elder@phs.scot

## 5. Configuring your RStudio Pro environment

### Global Options

- Once your RStudio Pro session has started, please ensure that you configure your settings as follows:
  - Access the RStudio Global Options menu by going to _Tools > Global Options..._
  - In the General menu, be sure that the settings match those in the screenshot below:

![Global Options](https://github.com/user-attachments/assets/c1441dba-b6af-4bec-b017-cc371bd14091)

  - _Optional, but recommended:_ On the Appearance menu, change the editor font to "Cascadia Code".  This font has support for ligatures that map multiple characters to single glyphs e.g. rendering "<-" as a single arrow glyph.

![Cascadia Code](https://github.com/user-attachments/assets/591ad6c5-4fad-4c25-9142-4ae0df288428)

### Installing Packages

- You will need to install R packages that you need.  These will not transfer over from the old Posit environment.  In the old Posit environment, you can run the following R code to write out a CSV file listing the packages you had installed.  This might be helpful for re-installing the packages you need:

```r

> # Function to check and install rstudioapi if necessary
> install_rstudioapi <- function() {
>   if (!requireNamespace("rstudioapi", quietly = TRUE)) {
>     cat("The 'rstudioapi' package is not installed. Installing it now...\n")
>     install.packages("rstudioapi", quiet = TRUE)
>     if (!requireNamespace("rstudioapi", quietly = TRUE)) {
>       stop("Failed to install 'rstudioapi'. Please install it manually.")
>     }
>     cat("'rstudioapi' package has been successfully installed.\n")
>   }
> }
> 
> # Install rstudioapi if not present
> install_rstudioapi()

Reviewer: I'm certain this first bit isn't required, as `{rstudioapi}` will be installed by default on any RStudio instance.
```suggestion
```

Also, it would be useful to have instructions on how to use the saved CSV file to restore/install packages.

Author: It's intended that this R code is run on the old Posit environment.  `{rstudioapi}` is not installed by default on the old Posit environment.  Users need to install themselves first.  I'll leave this bit in.

Author: Just to say, we will not have time / resource to implement and test a solution for taking the exported CSV file with the list of R packages, and automate the installation of this list in the new Posit Workbench environment.
